### DISCUSSION TITLE
Is it possible to detect connected devices in Wifi AP mode?

### DISCUSSION THREAD

**Discussion by stc1988** [Moddable-OpenSource/moddable#1342] [Q&A] [answered]
I am trying to use QR codes for Wi-Fi configuration by applying [webConfigWifi](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/network/wifi/webConfigWifi) application.

My plan is adding processing for when a smartphone connects to the MCU in AP mode, but there is no way to detect connected devices.

**Comment by phoddie** [comment, marked as answer]
The WiFi class callback on ESP32 provides two callbacks related to this:

- "station_connect" for when a new connection to the AP is established (ESP-IDF `WIFI_EVENT_AP_STACONNECTED` message)
- "station_disconnect" for when a connection to the AP is lost  (ESP-IDF `WIFI_EVENT_AP_STADISCONNECTED` message)

I'm not sure about the details of integrating this into `webConfigWifi` but these messages could be used to achieve the result you describe.

**Comment by stc1988** [comment]
Thank you for reply.

This code works to monitor msg `station_connect` and  `station_disconnect `
``` js
    WiFi.accessPoint({
      ssid: AP_NAME,
      password: AP_PASSWORD,
    });

    let monitor = new WiFi(null, msg => {
      trace(`${msg}\n`)
    })
```

**Reply by phoddie** [reply]
Nice. That works because you can  have more than one Wi-Fi monitor instance active at the same time. That avoids a conflict with webConfigWifi (which also has a Wi-Fi instance).

**Comment by stc1988** [comment]
This is short demo.
https://x.com/stc1988/status/1787422792687780059

After connect to access point, screen shows next QR code that open configuration URL.

**Reply by phoddie** [reply]
Very cool. Thanks for sharing the video.
